### PR TITLE
SQS Integration (Aws Sdk)

### PR DESCRIPTION
**Pull request by colin-higgins**
Changes proposed in this pull request:
 - [x] Introduce SQS client instrumentation from the AWS SDK
 - [x] Tests verifying sample app sends spans
 - [x] Test framework refactors

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by bobuva** [comment]
@colin-higgins For the purpose of code review, it would be good to know the steps you went through to retrieve and set up the AWS SDK. One thing, for example, I'd like to do is verify that the integration points are complete, and being able to look at the AWS  SDK assemblies would allow me to do that.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+            var tracer = Tracer.Instance;
+            Scope scope = null;
+
+            try
+            {
+                var serviceName = $"{tracer.DefaultServiceName}-{ServiceName}";
+                scope = Tracer.Instance.StartActive(OperationName, serviceName: serviceName);
+                var span = scope.Span;
+                span.Type = SpanTypes.Http; // TODO, is this right?
```
Comment (bobuva) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: Since Amazon SQS is a queuing system, it seems that we should have a new `SpanType.Queue`.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    /// Tracing integration for AWSSDK.SQS.
+    /// </summary>
+    public static class AmazonSqsIntegration
+    {
+        private const string IntegrationName = "AWS";
+        private const string OperationName = "aws.command";
+        private const string AgentName = "dotnet-aws-sdk";
+
+        private const string Major3 = "3";
+        private const string Major3Minor3 = "3.3";
```
Comment (bobuva) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: I would rename these entries to `MinimumSupportedVersion_Major` and `MinimumSupportedVersion_Minor` or something like that.

Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: This is the format we're using for all the integrations. I'd prefer to keep it the same unless we decide to standardize on another format.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 1 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: I would agree if this code were acting as the queue, but it is only the client for the queue.

Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: This did make me notice that our operation name seems to differ from java's implementation.
https://github.com/DataDog/dd-trace-java/blob/00fe40f1fba5c0df65a8af80430feb2a3cba6eea/dd-java-agent/instrumentation/aws-java-sdk-2.2/src/test/groovy/AwsClientTest.groovy#L78

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 2 above)
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: That makes no sense to me. The symbolic name should represent what the intent is. Anyway, that's fine, maybe I'll change them all in a separate PR :)

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 1 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: In fact they seem to have a mapping of operation names: https://github.com/DataDog/dd-trace-java/blob/42b88c7b8ec99b8b9ac2a4d4dda90ebbbe9e5921/dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java#L37

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: This seems like a consistency issue then. I would check for RFCs that apply to naming operations if such an RFC exists. And maybe asking in the `apm-integrations` Slack channel.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 2 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: A little knot in that might be that we may support several version ranges within one integration.
The names might get excessively verbose.
It's fine with me though.

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: To me, having `Major4 = "4"` just gives you another place to make a mistake. Copy and paste and you get `Major4 = "5"`.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 1 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: I worked with @tylerbenson to get a better idea of what Java is doing.

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: Sounds good, thanks @colin-higgins !

File: PackageVersionsGeneratorDefinitions.json [hunk 3]
```diff
@@ -1,59 +1,67 @@
...
+    "SampleProjectName": "Samples.AspNetCoreMvc2",
+    "SampleTargetFramework": "netcoreapp2.1",
+    "NugetPackageSearchName": "Microsoft.AspNetCore.Mvc",
+    "MinVersion": "2.0.0",
+    "MaxVersionExclusive": "3.0.0"
+  },
+  {
+    "IntegrationName": "AmazonSqs",
+    "SampleProjectName": "Samples.Amazon.SQS",
+    "SampleTargetFramework": "netcoreapp2.1",
```
Comment (zacharycmontoya) [inline PackageVersionsGeneratorDefinitions.json]: Can you remove this `SampleTargetFramework` property? This will limit it to only running on 2.1 and not 3.0

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs (hunk 2 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs]: Bob makes some good points, I think this worth revisiting, though it doesn't necessarily need to be addressed in this PR

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersions.g.cs [hunk 4]
```diff
@@ -278,5 +278,219 @@ public class PackageVersions
...
 #endif
 
+        public static IEnumerable<object[]> AmazonSqs =>
+
+            new List<object[]>
+            {
+#if DEFAULT_SAMPLES
+                new object[] { string.Empty },
+#else
+                new object[] { "3.1.0" },
```
Comment (colin-higgins) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersions.g.cs]: This list is another reason I want to limit how many packages we test. This is a little ridiculous.

File: samples/Samples.Amazon.SQS/Samples.Amazon.SQS.csproj [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ApiVersion Condition="'$(ApiVersion)' == ''">3.3.3</ApiVersion>
+    <DefineConstants Condition="'$(ApiVersion)'>='3.0.0'">$(DefineConstants);SQS_3_0</DefineConstants>
```
Comment (zacharycmontoya) [inline samples/Samples.Amazon.SQS/Samples.Amazon.SQS.csproj]: Was there any breaking API change that required a separate `SQS_3_0` compilation flag?

File: PackageVersions.g.props [hunk 6]
```diff
@@ -553,5 +553,617 @@ NOTE:  This code was generated by the GeneratePackageVersions tool. To safely
       <Properties>ApiVersion=2.2.0;RestoreRecursive=false;BuildProjectReferences=false</Properties>
       <TargetFramework>netcoreapp2.1</TargetFramework>
     </PackageVersionSample>
+    <PackageVersionSample Include="samples*\**\Samples.Amazon.SQS.csproj">
```
Comment (zacharycmontoya) [inline PackageVersions.g.props]: Yeah this is a massive list of versions. Maybe we should prioritize your major.minor testing change before merging this 😅

File: samples/Samples.Amazon.SQS/Samples.Amazon.SQS.csproj (hunk 5 above)
Comment (colin-higgins) [reply samples/Samples.Amazon.SQS/Samples.Amazon.SQS.csproj]: Ah, nope, I was just playing.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs [hunk 7]
```diff
@@ -115,7 +115,13 @@ public static object GetResponse(object webRequest, int opCode, int mdToken, lon
...
         /// <returns>Returns the value returned by the inner method call.</returns>
         [InterceptMethod(
-            TargetAssembly = "System.Net",
+            TargetAssembly = "System", // .NET Framework
+            TargetType = WebRequestTypeName,
+            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Net.WebResponse>" },
+            TargetMinimumVersion = Major4,
+            TargetMaximumVersion = Major4)]
+        [InterceptMethod(
+            TargetAssembly = "System.Net.Requests", // .NET Core
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Maybe this change was already in the branch when I handed it off, but is this change necessary for this PR or can it be moved out?

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+namespace Datadog.Trace.ClrProfiler.Integrations
+{
+    internal static class AwsSdkTags
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs]: Would you be opposed to renaming the class `AmazonSdkTags`? I know the tags in this class all start with `aws` but it seems a little confusing that we have different prefixes for filenames in this directory.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersions.g.cs (hunk 4 above)
Comment (bobuva) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersions.g.cs]: see what you mean

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs (hunk 8 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs]: Maybe use "aws" for all of them instead of "amazon"?

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs (hunk 7 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: The change from `System.Net` to `System` seems weird. Was this integration not working on .NET Framework?

File: samples/Samples.Amazon.SQS/Properties/launchSettings.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
+  "profiles": {
+    "Samples.AwsSqs": {
```
Comment (zacharycmontoya) [inline samples/Samples.Amazon.SQS/Properties/launchSettings.json]: nit: The name of this profile doesn't match the name of the sample program `Samples.Amazon.SQS`

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Datadog.Trace.ClrProfiler.IntegrationTests.csproj [hunk 10]
```diff
@@ -6,14 +6,8 @@
...
-    <None Include="..\..\src\Datadog.Trace.ClrProfiler.Native\bin\$(Configuration)\$(Platform)\**"
-             CopyToOutputDirectory="Always"
-             CopyToPublishDirectory="Always"
-             Link="profiler-lib\%(RecursiveDir)\%(Filename)%(Extension)" />
-    <Content Include="..\..\integrations.json"
-             CopyToOutputDirectory="Always"
-             CopyToPublishDirectory="Always"
-             Link="profiler-lib\integrations.json" />
+    <None Include="..\..\src\Datadog.Trace.ClrProfiler.Native\bin\$(Configuration)\$(Platform)\**" CopyToOutputDirectory="Always" CopyToPublishDirectory="Always" Link="profiler-lib\%(RecursiveDir)\%(Filename)%(Extension)" />
+    <None Remove="..\..\src\Datadog.Trace.ClrProfiler.Native\bin\Debug\x64\**" />
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Datadog.Trace.ClrProfiler.IntegrationTests.csproj]: What is this change needed for?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            _expectations.Add(CreateExpectation("DeleteMessage"));
+            _expectations.Add(CreateExpectation("SendMessageBatch"));
+            _expectations.Add(CreateExpectation("SendMessageBatch"));
+            _expectations.Add(CreateExpectation("ReceiveMessage"));
+            _expectations.Add(CreateExpectation("DeleteMessageBatch"));
+            _expectations.Add(CreateExpectation("PurgeQueue"));
+            _expectations.Add(CreateExpectation("DeleteQueue"));
+        }
+
+#if NETCOREAPP2_1
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs]: Is there a reason this is conditional on netcoreapp2.1? I would imagine some of the expectations above might be, since the .NET Core API doesn't include synchronous operations.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        [Trait("Category", "EndToEnd")]
+        public void SubmitsTraces(string packageVersion)
+        {
+            int agentPort = TcpPortProvider.GetOpenPort();
+            using (var agent = new MockTracerAgent(agentPort))
+            using (var processResult = RunSampleAndWaitForExit(agent.Port, packageVersion: packageVersion))
+            {
+                Assert.True(processResult.ExitCode >= 0, $"Process exited with code {processResult.ExitCode} and exception: {processResult.StandardError}");
+
+                var expectedSpans = 13;
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs]:
```suggestion
                var expectedSpans = _expectations.Count;
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs (hunk 8 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs]: "aws" across the board makes sense to me, I'd support that change 👍

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                if (expectedValue != null && actualValue != expectedValue)
+                {
+                    return FailureMessage(name: keyForMessage, actual: actualValue, expected: expectedValue);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterTagExpectation(string key, string expected, Func<MockTracerAgent.Span, bool> when, Func<string, string> customMessage = null)
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Could you rename `customMessage` to `customErrorMessageDelegate` or something comparable? That would help me with understanding the code more easily

Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Also, is `customMessage` used anywhere? Or is that just added more future extensibility?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+        protected string FailureMessage(string name, string actual, string expected)
+        {
+            return $"({name} mismatch: actual: {actual ?? "NULL"}, expected: {expected ?? "NULL"})";
+        }
+
+        protected string GetTag(MockTracerAgent.Span span, string tag)
+        {
+            if (span.Tags.ContainsKey(tag))
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: nit: This doesn't need any changes, but I always support using `TryGetValue` when possible, which could condense these lines to:

```
span.Tags.TryGetValue(tag, out string value);
return value;
```

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+                if (failures.Any())
+                {
+                    return string.Join(",", failures);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterCustomExpectation(string keyForMessage, Func<MockTracerAgent.Span, string> actual, Func<MockTracerAgent.Span, string> expected)
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Could we simplify this so that the `expected` param is not a Func but instead just a string that we compare against?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+            // Expectations for all spans regardless of type should go here
+            RegisterDelegateExpectation(ExpectBasicSpanDataExists);
+
+            RegisterCustomExpectation(nameof(OperationName), actual: s => s.Name, expected => OperationName);
+            RegisterCustomExpectation(nameof(ServiceName), actual: s => s.Service, expected => ServiceName);
+            RegisterCustomExpectation(nameof(Type), actual: s => s.Type, expected => Type);
+            RegisterCustomExpectation(nameof(ResourceName), actual: s => s.Resource.TrimEnd(), expected => ResourceName);
+
+            RegisterTagExpectation(
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: This seems to be the only `RegisterTagExpectation` that uses the `when` parameter. What are your thoughts on changing this semantic check to using the `RegisterDelegateExpectations` method instead and simplifying `RegisterTagExpectation` to not have the additional `when` parameter?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/GraphQLSpanExpectation.cs [hunk 17]
```diff
@@ -17,70 +23,22 @@ public class GraphQLSpanExpectation : WebServerSpanExpectation
...
-                mismatches.Add(FailureMessage(nameof(ServiceName), actual: span.Service, expected: ServiceName));
-            }
-
-            if (span.Type != Type)
+            if (string.IsNullOrEmpty(GetTag(span, Tags.ErrorMsg)))
             {
-                mismatches.Add(FailureMessage(nameof(Type), actual: span.Type, expected: Type));
+                if (IsGraphQLError)
+                {
+                    yield return "Expected an error message.";
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/GraphQLSpanExpectation.cs]:
```suggestion
                    yield return "Expected an error message but Tags.ErrorMsg was empty.";
```

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/GraphQLSpanExpectation.cs [hunk 18]
```diff
@@ -17,70 +23,22 @@ public class GraphQLSpanExpectation : WebServerSpanExpectation
...
-
-            var spanIsError = !string.IsNullOrEmpty(GetTag(span, Tags.ErrorMsg));
-
-            if (spanIsError != IsGraphQLError)
+            else
             {
-                mismatches.Add(FailureMessage(nameof(IsGraphQLError), actual: spanIsError.ToString(), expected: IsGraphQLError.ToString()));
+                if (!IsGraphQLError)
+                {
+                    yield return "Expected no error message.";
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/GraphQLSpanExpectation.cs]:
```suggestion
                    yield return $"Expected no error message but Tags.ErrorMsg was {GetTag(span, Tags.ErrorMsg)}.";
```

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs [hunk 19]
```diff
@@ -20,34 +21,36 @@ public class WebServerTestHelpers
...
-                    failures.Add($"{expectation.ResourceName} failed with: {failureMessage}");
+                    failures.Add($"{detail} failed with: {failureMessage}");
                 }
             }
 
             var finalMessage = Environment.NewLine + string.Join(Environment.NewLine, failures.Select(f => " - " + f));
 
             Assert.True(!failures.Any(), finalMessage);
+
+            var uninspectedSpans = spans.Count(s => s.Inspected);
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: I think this is supposed to be:
```suggestion
            var uninspectedSpans = spans.Count(s => !s.Inspected);
```

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs [hunk 20]
```diff
@@ -6,11 +6,12 @@
 
 namespace Datadog.Trace.ClrProfiler.IntegrationTests
 {
-    public class WebServerTestHelpers
+    public class SpanTestHelpers
     {
-        public static void AssertExpectationsMet(
-            List<WebServerSpanExpectation> expectations,
+        public static void AssertExpectationsMet<T>(
+            List<T> expectations,
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: nit: Could this just be `List<SpanExpectation>`?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+        {
+            return $"service={ServiceName}, operation={OperationName}, type={Type}, resource={ResourceName}";
+        }
+
+        /// <summary>
+        /// Override for custom filters.
+        /// </summary>
+        /// <param name="span">The span on which to filter.</param>
+        /// <returns>Whether the span qualifies for this expectation.</returns>
+        public virtual bool ShouldInspect(MockTracerAgent.Span span)
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: I feel like another verb would help make this a little bit clearer when reading over SpanTestHelpers.cs. Maybe `MatchesSpan`?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 14 above)
Comment (lucaspimentel) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: We also have a helper `GetValueOrDefault()` extension method if you don't like `out` parameters.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs (hunk 7 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: It was already in the branch, yes. Let's create a card to follow up on it?

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs (hunk 8 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs]: I'm fine with that, the services in Amazon are listed as `Amazon.Sqs`, `Amazon.DynamoDB`, etc.
Amazon just... spoke nicely for me. Rather than seeing a giant acronym for everything.
Either way, thoughts?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Datadog.Trace.ClrProfiler.IntegrationTests.csproj (hunk 10 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Datadog.Trace.ClrProfiler.IntegrationTests.csproj]: Not actually sure what happened here... guessing it was an issue when I rebased and conflicts got resolved?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 16 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: I'm fine with that, but I do like it being explicit. I think it forces the person writing the test to think about *when* tags apply and when they don't. Thoughts?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 21 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: How about just `Matches(span)`?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 15 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Sure, we can always change it back if we need to run actual logic, or I suppose the delegate one could be used.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 13 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: It was used before, I can delete it.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs (hunk 20 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: It could, but then every List of expectations would need to be an explicit list of `SpanExpectation`  and not the implementation. It's workable obviously, but it was less burden on the test class IMO. 
That way the list matches the implementation for that test and any extra properties or methods or whatever are just readily available to the test class.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs (hunk 11 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/AWS/AmazonSQSTests.cs]: The test now runs for framework

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 21 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Yeah that sounds good to me 👍

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 16 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: Yeah that's a good point. Can either have two overloads or make `when` a default argument that defaults to `Always`? That way we don't have to specify it everywhere

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs (hunk 20 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: Hmm. Maybe I'm missing something, but I don't see what's wrong with making all of the lists be of the parent class `SpanExpectation` since this method is only operating on that API level. This is a small nit so I don't want to hold up the review on it, but if my thinking is correct, this could help reinforce a separation of concerns by making this helper always use the overarching `SpanExpectation` API and it's the responsibility of the subclasses to implement them correctly.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs (hunk 7 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Oh wait, this was intended to be here!! I vaguely remember that this identified an outgoing http call made by the AWS sdk.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Yeah I think this was not working on .NET Framework. Colin, since you're actively working on this branch, could you verify that removing this integration causes no "http" spans to show up on the .NET Framework calls? There should basically be an "aws" span wrapping an "http" span for each call

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs [hunk 22]
```diff
@@ -6,11 +6,12 @@
...
             List<MockTracerAgent.Span> spans)
+            where T : SpanExpectation
```
Comment (colin-higgins) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: This constraint does force them to be `SpanExpectation` @zacharycmontoya 
Does this not accomplish what you're thinking about? https://github.com/DataDog/dd-trace-dotnet/pull/546/files/593302562e75915a279de2fd2720f2d285be3d95#r343942759

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs (hunk 8 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AwsSdkTags.cs]: Amazon has a better ring to it, but I feel like we'll always be referring to these integrations as `aws` like "Hey we gotta add more to that AWS integration". So my preference would be to name things `aws` as much as possible. Like change everything from `Amazon` to `aws` and then if something sounds ridiculous we can roll it back

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs (hunk 16 above)
Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs]: sure

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs (hunk 22 above)
Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanTestHelpers.cs]: You're right, it does accomplish what I'm thinking about. I guess I'm just being nitpicky in that I'd prefer strongly typing this to `SpanExpectation` than using generics and this constraint, but it's fine the way it is so I'm perfectly okay with this going in. Just expressing a coding preference 😄

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs (hunk 7 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: My concern is that by changing this from `System.Net` to `System` we will no longer instrument outgoing http requests in systems where `WebRequest` lives in `System.Net.dll`. Should we add both to cover all the bases?

Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: I'm totally happy to have both. What do you think?

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: I don't think we need `System.Net`. This seems like the right fix, thanks. (resolving)

File: test/Datadog.Trace.TestHelpers/MockTracerAgent.cs [hunk 23]
```diff
@@ -146,42 +146,6 @@ protected virtual void OnRequestReceived(HttpListenerContext context)
             RequestReceived?.Invoke(this, new EventArgs<HttpListenerContext>(context));
         }
 
-        private static List<Span> ToSpans(dynamic data)
```
Comment (lucaspimentel) [inline test/Datadog.Trace.TestHelpers/MockTracerAgent.cs]: Should we merge the PR that this come from first so this (unrelated) diff is not here?

Comment (lucaspimentel) [inline test/Datadog.Trace.TestHelpers/MockTracerAgent.cs]: Should we merge the PR that this come from first so this (unrelated) diff is not here?

Comment (colin-higgins) [reply test/Datadog.Trace.TestHelpers/MockTracerAgent.cs]: I'm fine with that.

Comment (lucaspimentel) [reply test/Datadog.Trace.TestHelpers/MockTracerAgent.cs]: https://github.com/DataDog/dd-trace-dotnet/pull/501

**Comment by lucaspimentel** [comment]
Closing this PR for now. We will refactor this integration to use the new "call target" instrumentation in the near future.
